src/runtime-rs/crates/hypervisor/src/device/pci_path.rs

```diff
@@ -0,0 +1,259 @@
+    type Error = anyhow::Error;
+
+    fn try_from(v: u32) -> Result<PciSlot> {
+        if v > MAX_PCI_SLOTS {
+            return Err(anyhow!("value {:?} exceeds MAX: {:?}", v, MAX_PCI_SLOTS));
+        }
+
+        Ok(PciSlot(v as u8))
+    }
+}
+
+// A PciPath describes where a PCI sits in a PCI hierarchy.
+//
+// Consists of a list of PCI slots, giving the slot of each bridge
+// that must be traversed from the PCI root to reach the device,
+// followed by the slot of the device itself.
+//
+// When formatted into a string is written as "xx/.../yy/zz". Here,
+// zz is the slot of the device on its PCI bridge, yy is the slot of
+// the bridge on its parent bridge and so forth until xx is the slot
+// of the "most upstream" bridge on the root bus.
+//
+// If a device is directly connected to the root bus, which used in
+// lightweight hypervisors, such as dragonball/firecracker/clh, and
+// its PciPath.slots will contains only one PciSlot.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct PciPath {
+    // list of PCI slots
+    pub slots: Vec<PciSlot>,
+}
+
+impl PciPath {
+    pub fn new(slots: Vec<PciSlot>) -> Option<PciPath> {
+        if slots.is_empty() {
+            return None;
+        }
+
+        Some(PciPath { slots })
+    }
+
+    // device_slot to get the slot of the device on its PCI bridge
+    pub fn get_device_slot(&self) -> Option<PciSlot> {
+        self.slots.last().cloned()
+    }
+
+    // root_slot to get the slot of the "most upstream" bridge on the root bus
+    pub fn get_root_slot(&self) -> Option<PciSlot> {
+        self.slots.first().cloned()
+    }
+}
+
+impl ToString for PciPath {
+    // method to format the PciPath into a string
+    fn to_string(&self) -> String {
+        self.slots
+            .iter()
+            .map(|pci_slot| format!("{:02x}", pci_slot.0))
+            .collect::<Vec<String>>()
+            .join("/")
+    }
+}
+
+// convert from u32
+impl TryFrom<u32> for PciPath {
+    type Error = anyhow::Error;
+
+    fn try_from(slot: u32) -> Result<PciPath> {
+        Ok(PciPath {
+            slots: vec![PciSlot::try_from(slot).context("pci slot convert failed.")?],
+        })
+    }
+}
+
+impl TryFrom<&str> for PciPath {
+    type Error = anyhow::Error;
+
+    // method to parse a PciPath from a string
+    fn try_from(path: &str) -> Result<PciPath> {
+        if path.is_empty() {
+            return Err(anyhow!("path given is empty."));
+        }
+
+        let mut pci_slots: Vec<PciSlot> = Vec::new();
+        let slots: Vec<&str> = path.split('/').collect();
+        for slot in slots {
+            match PciSlot::try_from(slot) {
+                Ok(s) => pci_slots.push(s),
+                Err(e) => return Err(anyhow!("slot is invalid with: {:?}", e)),
+            }
+        }
+
+        Ok(PciPath { slots: pci_slots })
+    }
+}
+
+// get pci path from bdf
+pub fn get_pcipath_from_bdf(bdf: &str) -> Option<u32> {
+    if !bdf.starts_with("0000:00") {
```

Comment:
VMM that supports complex PCI hierarchy may not start with "0000:00" in BDF. 
In that case, they could not get pci path through this function, is that expected?

Reply:
Fixed. It was useless now, remove it.